---
title: "Data Retrieval from TCGA (BRCA)"
output: html_document
author: "Chaitra Rao"
---

# Data Retrieval
Retrieval of the following data:

* 1. RNA Seq gene expression with normalized counts
* 2. DNA Methylation profiles obtained via HumanMethylation450 BeadChip (probes for ca. 450K CpG sites)
* 3. Somatic non-silent mutations
* 4. CNA data profiled by SNP-arrays  - log2ratios of copy number changes between samples to refernce samples


```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE, cache = TRUE, cache.lazy = FALSE)
```


## Necessary Packages

```{r libraries, warning=FALSE, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
suppressPackageStartupMessages(library(MOFA))
suppressPackageStartupMessages(library(MultiAssayExperiment))
```

## Retrieval of data

```{r BRCA_data}
rnaseq.BRCA <- getTCGA(disease = "BRCA", data.type = "RNASeq2", type = "RPKM")
meth.BRCA <- getTCGA(disease = "BRCA", data.type = "Methylation", type = "450K")
mut.BRCA <- getTCGA(disease = "BRCA", data.type = "Mutation", type = "somatic")
#cna.BRCA <- curatedTCGAData(diseaseCode = "BRCA",
                        #assays = "*GISTIC*",
                        #dry.run = FALSE)
cna.BRCA <- getTCGA(disease="BRCA", data.type="CNA_SNP", clinical = TRUE)

```

### Downsize all the data to top 1% variance

```{r meth_downsize}
meth.var <- apply(meth.BRCA$dat, 1, function(x) var(x, na.rm=TRUE))
meth.data.BRCA <- subset(meth.BRCA$dat, meth.var >= quantile(meth.var, 0.90, na.rm = TRUE))
```

```{r rnaseq_downsize}
rnaseq.var <- apply(rnaseq.BRCA$dat, 1, function(x) var(x, na.rm=TRUE))
rnaseq.data.BRCA <- subset(rnaseq.BRCA$dat, rnaseq.var >= quantile(rnaseq.var, 0.90, na.rm=TRUE))
```

```{r mut_downsize}
mut.var <- apply(mut.BRCA$dat, 1, function(x) var(x, na.rm=TRUE))
mut.data.BRCA <- subset(mut.BRCA$dat, mut.var >= quantile(mut.var, 0.90, na.rm=TRUE))
```

```{r cna_downsize}
cna.var <- apply(cna.BRCA$dat, 1, function(x) var(x, na.rm=TRUE))
cna.data.BRCA <- subset(cna.BRCA$dat, cna.var >= quantile(cna.var, 0.90, na.rm=TRUE))
```

### Take delta values for methylation and RNAseq (tumor - normal)

```{r delta_meth}
meth.BRCA.tum.norm <- TumorNormalMatch(meth.data.BRCA)
meth_tum <- (meth.BRCA.tum.norm$primary)
meth_norm <- (meth.BRCA.tum.norm$normal)
delta_meth <- (meth_tum - meth_norm)
```

```{r delta_rnaseq}
rnaseq.BRCA.tum.norm <- TumorNormalMatch(rnaseq.data.BRCA)
rnaseq_tum <- (rnaseq.BRCA.tum.norm$primary)
rnaseq_norm <- (rnaseq.BRCA.tum.norm$normal)
delta_rnaseq <- (rnaseq_tum - rnaseq_norm)
```

### Convert beta to M values

```{r beta2m}
delta_meth <- beta2m(delta_meth)
```

### RemBRCAe NA values from Methylation data

```{r remBRCAe_NA}
delta_meth <- na.omit(delta_meth)
```


### Clinical data

```{r store_clinical}
clin_data.BRCA <- data.matrix(cna.BRCA$clinical)
```

### Filtering tumor CNA

CNA downloaded this way has both normal and tumor readouts; getting rid of normal tissue (that is, getting barcodes with `01` as fourth element at the barcode).

```{r filter_tumor_cna}
is_tumor <- sapply(strsplit(colnames(cna.data.BRCA), '-'), function(x) return(grepl('^01', x[4])))
cna_tumors <- cna.data.BRCA[,colnames(cna.data.BRCA)[is_tumor]]
```

### Trimming the colnames (TCGA barcodes)

```{r trimming}
string1 <- colnames(delta_rnaseq)
colnames(delta_rnaseq) <- substring(string1, 1, 12)
string2 <- colnames(delta_meth)
colnames(delta_meth) <- substring(string2, 1, 12)
string3 <- colnames(mut.data.BRCA)
colnames(mut.data.BRCA) <- substring(string3, 1, 12)
## string4 <- colnames(cna.BRCA$dat)
## colnames(cna.BRCA$dat) <- substring(string4, 1, 12)
colnames(cna_tumors) <-  substring(colnames(cna_tumors), 1, 12)

```


```{r datamatrices}
rnaseq_data <- delta_rnaseq
meth_data <- delta_meth
mut_data <- mut.data.BRCA
cna_data <- cna_tumors

```


### Making a list out of the retrieved data and getting the coldata for MOFA

```{r mofalist}
BRCA_list <- list(rnaseq = rnaseq_data,
          meth = meth_data,
          mut = mut_data,
          cna = cna_data)

#Ensuring duplicate colnames are not present, and taking them out if so
sapply(BRCA_list, function(x) table(duplicated(colnames(x))))
## these are duplicated
sapply(BRCA_list, function(x) colnames(x)[duplicated(colnames(x))])
## not going further on the reasons of the duplications: just getting unique stuff
for (item in names(BRCA_list)) {
    BRCA_list[[item]] <- BRCA_list[[item]][,unique(colnames(BRCA_list[[item]]))]
}
sapply(BRCA_list, function(x) table(duplicated(colnames(x))))
```

```{r columndata}
cdata = as.data.frame(clin_data.BRCA)
```

## Getting the MOFAobject

```{r mofaobject}
BRCA_mae <- MultiAssayExperiment(
  experiments = BRCA_list, 
  colData = cdata)

BRCA_data <- createMOFAobject(BRCA_mae)
```

# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```



