
```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE)
```

---
title: "Data Retrieval from TCGA"
output: html_document
author: "Chaitra Rao"
---

## Necessary Packages

```{r libararies, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
```

## Retrieval of data

Retrieval of the following data:
1. RNA Seq gene expression with raw counts
2. DNA Methylation profiles obtained via HumanMethylation450 BeadChip (probes for ca. 450K CpG sites)
3. Somatic non-silent mutations
4. GISTIC-called CNA data summarized by gene -> the quantitative readout can be discreticized -> Ref. to the datatable


```{r COAD data, echo=FALSE}

rnaseq.COAD <- getTCGA(disease = "COAD", data.type = "RNASeq2", type = "count", clinical = TRUE)
meth.COAD <- getTCGA(disease = "COAD", data.type = "Methylation", type = "450K", clinical = TRUE)
mut.COAD <- getTCGA(disease = "COAD", data.type = "Mutation", type = "somatic")
cna.COAD <- curatedTCGAData(diseaseCode = "COAD",
                        assays = "*GISTIC*",
                        dry.run = FALSE)

```

## Getting a datatable for each kind of data

Storing the data matrices in a datatable.

```{r datatable, echo=FALSE}
data.matrix(rnaseq.COAD$dat)
meth_data <- data.matrix(meth.COAD$dat)
mut_dat <- data.matrix(mut.COAD$dat)
cna_data <- data.matrix(assays(cna.COAD)[[2]])
```


Store the clinical data in a datatable

```{r store clinical, echo=FALSE}
clin_data <- data.matrix(rnaseq.COAD$clinical)
```


##Get matched tumor and normal samples

```{r get matched tumor/normal smaples, echo = FALSE}
rnaseq.COAD.tum.norm <- TumorNormalMatch(rnaseq.COAD$dat)
meth.COAD.tum.norm <- TumorNormalMatch(meth.COAD$dat)
mut.COAD.tum.norm <- TumorNormalMatch(mut.COAD$dat)
cna.COAD.tum.norm <- TumorNormalMatch(assays(cna.COAD)[[2]])

```


Store the tumor and normal sample in two separate datatables (dim = gene x sample)

RNASeq
```{r store rnaseq matched, echo=FALSE}
rnaseq_tum <- data.matrix(rnaseq.COAD.tum.norm$primary)
rnaseq_norm <- data.matrix(rnaseq.COAD.tum.norm$normal)

```

Methylation
```{r store meth matched, echo=FALSE}
meth_tum <- data.matrix(meth.COAD.tum.norm$primary)
meth_norm <- data.matrix(meth.COAD.tum.norm$normal)

```

Mutation
```{r store mut matched, echo=FALSE}
mut_tum <- data.matrix(mut.COAD.tum.norm$primary)
mut_norm <- data.matrix(mut.COAD.tum.norm$normal)

```

CNA
```{r store cna matched, echo=FALSE}
cna_tum <- data.matrix(cna.COAD.tum.norm$primary)
cna_norm <- data.matrix(cna.COAD.tum.norm$normal)

```

# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```


#data matrices instead of datatables without annotations
#

