---
title: "Data Retrieval from TCGA (OV)"
output: pdf_document
author: "Chaitra Rao"
---

```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE, cache = TRUE, cache.lazy = FALSE)
```

## Necessary Packages

```{r libararies, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
suppressPackageStartupMessages(library(MOFA))
suppressPackageStartupMessages(library(MultiAssayExperiment))
```

## Retrieval of data

Retrieval of the following data from the TCGA database:
1. RNA Seq gene expression with raw counts
2. DNA Methylation profiles obtained via HumanMethylation450 BeadChip (probes for ca. 450K CpG sites)
3. Somatic non-silent mutations
4. GISTIC-called CNA data summarized by gene -> the quantitative readout can be discreticized

```{r LUAD_data, echo=FALSE}
rnaseq.OV <- getTCGA(disease = "OV", data.type = "RNASeq2", type = "count")
meth.OV <- getTCGA(disease = "OV", data.type = "Methylation", type = "450K")
mut.OV <- getTCGA(disease = "OV", data.type = "Mutation", type = "somatic")
cna.OV <- curatedTCGAData(diseaseCode = "OV",
                        assays = "*GISTIC*",
                        dry.run = FALSE)
cna_OV <- getTCGA(disease="OV", data.type="CNA_SNP", clinical = TRUE) #get the clinical data from here

```

## Downsize all the retrieved data and include features that are at the top 1% variance across patients

```{r meth_downsize, echo = TRUE}
meth.var <- apply(meth.OV$dat, 1, var)
meth.data.OV <- subset(meth.OV$dat, meth.var >= quantile(meth.var, 0.99, na.rm=T) & !is.na(meth.var))

```

```{r rnaseq_downsize, echo = TRUE}
rnaseq.var <- apply(rnaseq.OV$dat, 1, var)
rnaseq.data.OV <- subset(rnaseq.OV$dat, rnaseq.var >= quantile(rnaseq.var, 0.99, na.rm=T) & !is.na(rnaseq.var))

```

```{r mut_downsize, echo = TRUE}
mut.var <- apply(mut.OV$dat, 1, var)
mut.data.OV <- subset(mut.OV$dat, mut.var >= quantile(mut.var, 0.99, na.rm=T) & !is.na(mut.var))

```

```{r cna_downsize, echo=TRUE}
cna.var <- apply((assays(cna.OV)[[2]]), 1, var)
cna.data.OV <- subset((assays(cna.OV)[[2]]), cna.var >= quantile(cna.var, 0.99, na.rm=T) & !is.na(cna.var))

```


## Convert methylation beta values to M values 

```{r beta2m, echo=TRUE, message=FALSE}
meth.data.OV <- beta2m(meth.data.OV)
```

## Remove NA values from Methylation data

```{r remove_NA, echo=TRUE}
meth.data.OV <- na.omit(meth.data.OV)
```


## Store the clinical data as a datamatrix

```{r store_clinical, echo=TRUE}
clin_data.OV <- data.matrix(cna_OV$clinical)
```

## Getting a datatable for each kind of data

```{r datatable, echo=TRUE}
DT::datatable(head(rnaseq.OV$dat))
DT::datatable(head(meth.OV$dat))
DT::datatable(head(mut.OV$dat))
DT::datatable(head(assays(cna.OV)[[2]]))
```

CNA downloaded this way has both normal and tumor readouts; getting rid of normal tissue (that is, getting barcodes with `01` as fourth element at the barcode).

```{r filter_tumor_cna, echo=TRUE}
is_tumor <- sapply(strsplit(colnames(cna.data.OV), '-'), function(x) return(grepl('^01', x[4])))
cna_tumors <- cna.data.OV[,colnames(cna.data.OV)[is_tumor]]
```

## Trimming the colnames (TCGA barcodes)

```{r trimming, echo=TRUE, message=FALSE}
string1 <- colnames(rnaseq.data.OV)
colnames(rnaseq.data.OV) <- substring(string1, 1, 12)
string2 <- colnames(meth.data.OV)
colnames(meth.data.OV) <- substring(string2, 1, 12)
string3 <- colnames(mut.data.OV)
colnames(mut.data.OV) <- substring(string3, 1, 12)
## string4 <- colnames(cna.COAD$dat)
## colnames(cna.COAD$dat) <- substring(string4, 1, 12)
colnames(cna_tumors) <-  substring(colnames(cna_tumors), 1, 12)

```

## Replace all values of the CNA data <0 by -1 and >0 by 1

```{r merge_classes_cna, echo = TRUE}
cna_tumors_r <- replace(cna_tumors, cna_tumors < 0, -1)
cna_tumors_r <- replace(cna_tumors_r, cna_tumors_r > 0, 1)
```

## Store the data matrices in separate variables so that they are easily accessible later

```{r datamatrices, echo=TRUE}
rnaseq_data <- rnaseq.data.OV
meth_data <- meth.data.OV
mut_data <- mut.data.OV
## cna_data <- cna.COAD$dat
##cna_data <- data.matrix(assays(cna.COAD)[[2]])

```

## Making a list out of the data retrieved and getting the coldata for MOFA


```{r mofalist, echo = TRUE}
OV_list <- list(rnaseq = rnaseq_data,
          meth = meth_data,
          mut = mut_data,
          cna = cna_tumors_r)


#Ensuring duplicate colnames are not present, and taking them out if so
sapply(OV_list, function(x) table(duplicated(colnames(x))))
## these are duplicated
sapply(OV_list, function(x) colnames(x)[duplicated(colnames(x))])
## not going further on the reasons of the duplications: just getting unique stuff
for (item in names(OV_list)) {
    OV_list[[item]] <- OV_list[[item]][,unique(colnames(OV_list[[item]]))]
}
sapply(OV_list, function(x) table(duplicated(colnames(x))))
```


```{r columndata, echo = TRUE}
cdata = as.data.frame(clin_data.OV)
```

## Create an untrained MOFA object from a MultiAssayExperiment object

```{r mofaobject, echo = TRUE}
ov_mae <- MultiAssayExperiment(
  experiments = OV_list, 
  colData = cdata)

OV_data <- createMOFAobject(ov_mae)

```

# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```




