---
title: "Data Retrieval from TCGA (COAD)"
output: html_document
author: "Chaitra Rao"
---

```{r, cache = TRUE, echo=FALSE}
<<<<<<< HEAD
knitr::opts_chunk$set(error = TRUE, cache = TRUE, cache.lazy = FALSE)
=======
knitr::opts_chunk$set(error = TRUE, cache = TRUE)
>>>>>>> ae12ac69af0c09d932def3b3a48d8128136c5562
```

## Necessary Packages

```{r libraries, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
```

## Retrieval of data

Retrieval of the following data:
1. RNA Seq gene expression with raw counts
2. DNA Methylation profiles obtained via HumanMethylation450 BeadChip (probes for ca. 450K CpG sites)
3. Somatic non-silent mutations
4. GISTIC-called CNA data summarized by gene -> the quantitative readout can be discreticized -> Ref. to the datatable


```{r COAD data, echo=FALSE}
rnaseq.COAD <- getTCGA(disease = "COAD", data.type = "RNASeq2", type = "count", clinical = TRUE)
meth.COAD <- getTCGA(disease = "COAD", data.type = "Methylation", type = "450K", clinical = TRUE)
mut.COAD <- getTCGA(disease = "COAD", data.type = "Mutation", type = "somatic")
cna.COAD <- getTCGA(disease="COAD", data.type="CNA_SNP", clinical = TRUE)
#cna.COAD <- curatedTCGAData(diseaseCode = "COAD",
                        #assays = "*GISTIC*",
                       # dry.run = FALSE)
```

```{r beta2m, echo=FALSE, message=FALSE}
meth.COAD$dat <- beta2m(meth.COAD$dat)
```

Print the clinical data in a datamatrix

```{r store clinical, echo=FALSE}
clin_data <- data.matrix(cna.COAD$clinical)
```


```{r print_datatables, echo=FALSE, message=FALSE}
DT::datatable(head(rnaseq.COAD$dat))
DT::datatable(head(meth.COAD$dat))
DT::datatable(head(mut.COAD$dat))
DT::datatable(head(cna.COAD$dat))
```

CNA downloaded that way has both normal and tumor readouts; getting rid of normal tissue (that is, getting barcodes with `01` as fourth element at the barcode).

```{r filter_tumor_cna}
is_tumor <- sapply(strsplit(colnames(cna.COAD$dat), '-'), function(x) return(grepl('^01', x[4])))
cna_tumors <- cna.COAD$dat[,colnames(cna.COAD$dat)[is_tumor]]
```

##Get matched tumor and normal samples

```{r get_matched, echo = FALSE}
rnaseq.COAD.tum.norm <- TumorNormalMatch(rnaseq.COAD$dat)
meth.COAD.tum.norm <- TumorNormalMatch(meth.COAD$dat)
```

Print the tumor and normal sample in two separate datatables (dim = gene x sample)

RNASeq
```{r print_matched_rnaseq, echo=FALSE}
rnaseq_tum <- (rnaseq.COAD.tum.norm$primary)
rnaseq_norm <- (rnaseq.COAD.tum.norm$normal)
```

Methylation
```{r print_matched_meth, echo=FALSE}
meth_tum <- (meth.COAD.tum.norm$primary)
meth_norm <- (meth.COAD.tum.norm$normal)
```

```{r trimming, echo=FALSE, message=FALSE}
string1 <- colnames(rnaseq.COAD$dat)
colnames(rnaseq.COAD$dat) <- substring(string1, 1, 12)
string2 <- colnames(meth.COAD$dat)
colnames(meth.COAD$dat) <- substring(string2, 1, 12)
string3 <- colnames(mut.COAD$dat)
colnames(mut.COAD$dat) <- substring(string3, 1, 12)
## string4 <- colnames(cna.COAD$dat)
## colnames(cna.COAD$dat) <- substring(string4, 1, 12)
colnames(cna_tumors) <-  substring(colnames(cna_tumors), 1, 12)
<<<<<<< HEAD
=======

>>>>>>> ae12ac69af0c09d932def3b3a48d8128136c5562
```

## Data list for MOFA

Ensuring duplicate colnames are not present, and taking them out if so

```{r datamatrices, echo=FALSE}
rnaseq_data <- rnaseq.COAD$dat
meth_data <- meth.COAD$dat
mut_data <- mut.COAD$dat
## cna_data <- cna.COAD$dat
##cna_data <- data.matrix(assays(cna.COAD)[[2]])
<<<<<<< HEAD
```

```{r mofalist}
d <- list(rna = rnaseq_data,
          meth = meth_data,
          mut = mut_data,
          cna = cna_tumors)
sapply(d, function(x) table(duplicated(colnames(x))))
## these are duplicated
sapply(d, function(x) colnames(x)[duplicated(colnames(x))])
## not going further on the reasons of the duplications: just getting unique stuff
for (item in names(d)) {
    d[[item]] <- d[[item]][,unique(colnames(d[[item]]))]
}
sapply(d, function(x) table(duplicated(colnames(x))))
```

MOFA object (suggestion: better do it from a `MultiAssayExperiment`)
=======

```

```{r mof}
d <- list(rna = rnaseq_data,
          meth = meth_data,
          mut = mut_data,
          cna = cna_tumors)

sapply(d, function(x) table(duplicated(colnames(x))))

## these are duplicated
sapply(d, function(x) colnames(x)[duplicated(colnames(x))])

## not going further on the reasons of the duplications: just getting unique stuff
for (item in names(d)) {
    d[[item]] <- d[[item]][,unique(colnames(d[[item]]))]
}

sapply(d, function(x) table(duplicated(colnames(x))))

```

MOFA object (suggestion: better do it from a `MultiAssayExperiment`)

```{r mofaobject}
MOFAobject <- createMOFAobject(d)
MOFAobject
>>>>>>> ae12ac69af0c09d932def3b3a48d8128136c5562

```{r mofaobject}
MOFAobject <- createMOFAobject(d)
MOFAobject
```


# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```
