---
title: "Data Retrieval from TCGA (COAD)"
output: html_document
author: "Chaitra Rao"
---

```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE, cache = TRUE, cache.lazy = FALSE)
```

## Necessary Packages

```{r libraries, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
suppressPackageStartupMessages(library(MOFA))
suppressPackageStartupMessages(library(MultiAssayExperiment))
```

## Retrieval of data

Retrieval of the following data:
1. RNA Seq gene expression with raw counts
2. DNA Methylation profiles obtained via HumanMethylation450 BeadChip (probes for ca. 450K CpG sites)
3. Somatic non-silent mutations
4. GISTIC-called CNA data summarized by gene -> the quantitative readout can be discreticized -> Ref. to the datatable

```{r LUAD_data, echo=FALSE}
rnaseq.COAD <- getTCGA(disease = "COAD", data.type = "RNASeq2", type = "RPKM")
meth.COAD <- getTCGA(disease = "COAD", data.type = "Methylation", type = "450K")
mut.COAD <- getTCGA(disease = "COAD", data.type = "Mutation", type = "somatic")
cna.COAD <- curatedTCGAData(diseaseCode = "COAD",
                        assays = "*GISTIC*",
                        dry.run = FALSE)
cna_COAD <- getTCGA(disease="COAD", data.type="CNA_SNP", clinical = TRUE)

```

## Downsize methylation data, include CpG sites that are at the top 1% variance

```{r meth_downsize, echo = FALSE}
meth.var <- apply(meth.COAD$dat, 1, var)
meth.data.COAD <- subset(meth.COAD$dat, meth.var >= quantile(meth.var, 0.99, na.rm=T) & !is.na(meth.var))
```

```{r rnaseq_downsize, echo = FALSE}
rnaseq.var <- apply(rnaseq.COAD$dat, 1, var)
rnaseq.data.COAD <- subset(rnaseq.COAD$dat, rnaseq.var >= quantile(rnaseq.var, 0.99, na.rm=T) & !is.na(rnaseq.var))
```

```{r mut_downsize, echo = FALSE}
mut.var <- apply(mut.COAD$dat, 1, var)
mut.data.COAD <- subset(mut.COAD$dat, mut.var >= quantile(mut.var, 0.99, na.rm=T) & !is.na(mut.var))
```

```{r cna_downsize, echo=FALSE}
cna.var <- apply((assays(cna.COAD)[[2]]), 1, var)
cna.data.COAD <- subset((assays(cna.COAD)[[2]]), cna.var >= quantile(cna.var, 0.99, na.rm=T) & !is.na(cna.var))
```

## Convert beta to M values

```{r beta2m, echo=FALSE, message=FALSE}
meth.data.COAD <- beta2m(meth.data.COAD)
```

## Remove NA values from Methylation data

```{r remove_NA, echo=FALSE}
meth.data.COAD <- na.omit(meth.data.COAD)
```


## Print the clinical data in a datamatrix

```{r store_clinical, echo=FALSE}
clin_data.COAD <- data.matrix(cna_COAD$clinical)
```

## Getting a datatable for each kind of data

```{r datatable, echo=FALSE}
DT::datatable(head(rnaseq.COAD$dat))
DT::datatable(head(meth.COAD$dat))
DT::datatable(head(mut.COAD$dat))
DT::datatable(head(assays(cna.COAD)[[2]]))
```

CNA downloaded this way has both normal and tumor readouts; getting rid of normal tissue (that is, getting barcodes with `01` as fourth element at the barcode).

```{r filter_tumor_cna, echo=FALSE}
is_tumor <- sapply(strsplit(colnames(cna.data.COAD), '-'), function(x) return(grepl('^01', x[4])))
cna_tumors <- cna.data.COAD[,colnames(cna.data.COAD)[is_tumor]]
```

## Get matched tumor and normal samples

```{r get_matched, echo = FALSE}
rnaseq.COAD.tum.norm <- TumorNormalMatch(rnaseq.COAD$dat)
meth.COAD.tum.norm <- TumorNormalMatch(meth.COAD$dat)

```

## Print the tumor and normal sample in two separate datatables (dim = gene x sample)

## RNASeq

```{r print_matched_rnaseq, echo=FALSE}
rnaseq_tum <- (rnaseq.COAD.tum.norm$primary)
rnaseq_norm <- (rnaseq.COAD.tum.norm$normal)
```

## Methylation

```{r print_matched_meth, echo=FALSE}
meth_tum <- (meth.COAD.tum.norm$primary)
meth_norm <- (meth.COAD.tum.norm$normal)
```

## Trimming the colnames (TCGA barcodes)

```{r trimming, echo=FALSE, message=FALSE}
string1 <- colnames(rnaseq.data.COAD)
colnames(rnaseq.data.COAD) <- substring(string1, 1, 12)
string2 <- colnames(meth.data.COAD)
colnames(meth.data.COAD) <- substring(string2, 1, 12)
string3 <- colnames(mut.data.COAD)
colnames(mut.data.COAD) <- substring(string3, 1, 12)
## string4 <- colnames(cna.COAD$dat)
## colnames(cna.COAD$dat) <- substring(string4, 1, 12)
colnames(cna_tumors) <-  substring(colnames(cna_tumors), 1, 12)

```

## Create two matrices for CNA, one for gains and one for losses

```{r merge_classes_cna_gains, echo = FALSE}
cna_tumors_gain <- replace(cna_tumors, cna_tumors < 0, 0)
cna_tumors_gain <- replace(cna_tumors_gain, cna_tumors_gain > 0, 1)
```

```{r merge_classes_cna_losses, echo = FALSE}
cna_tumors_loss <- replace(cna_tumors, cna_tumors > 0, 0)
cna_tumors_loss <- replace(cna_tumors_loss, cna_tumors_loss < 0, 1)
```


```{r datamatrices, echo=FALSE}
rnaseq_data <- rnaseq.data.COAD
meth_data <- meth.data.COAD
mut_data <- mut.data.COAD
## cna_data <- cna.COAD$dat
##cna_data <- data.matrix(assays(cna.COAD)[[2]])

```

## Making a list out of the retrieved data and getting the coldata for MOFA

```{r mofalist, echo = FALSE}
LUAD_list <- list(rnaseq = rnaseq_data,
          meth = meth_data,
          mut = mut_data,
          cna_gain = cna_tumors_gain,
          cna_loss = cna_tumors_loss)

#Ensuring duplicate colnames are not present, and taking them out if so
sapply(COAD_list, function(x) table(duplicated(colnames(x))))
## these are duplicated
sapply(COAD_list, function(x) colnames(x)[duplicated(colnames(x))])
## not going further on the reasons of the duplications: just getting unique stuff
for (item in names(COAD_list)) {
    COAD_list[[item]] <- COAD_list[[item]][,unique(colnames(COAD_list[[item]]))]
}
sapply(COAD_list, function(x) table(duplicated(colnames(x))))
```

```{r columndata, echo = FALSE}
cdata = as.data.frame(clin_data.COAD)
```

## Getting the MOFAobject

```{r mofaobject, echo = FALSE}
luad_mae <- MultiAssayExperiment(
  experiments = COAD_list, 
  colData = cdata)

COAD_data <- createMOFAobject(coad_mae)

```

# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```




