---
title: "Application of MOFA on BRCA data"
output: html_document
author: "Chaitra Rao"
---

```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE, cache.lazy = FALSE)
```

## Necessary Packages

```{r libraries, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
suppressPackageStartupMessages(library(MOFA))
suppressPackageStartupMessages(library(MultiAssayExperiment))
suppressPackageStartupMessages(library(survival))
```

Saving the genomic information of all the CpG sites.
```{r}
genomic_info <- meth.BRCA$cpgs
```

Characterictics of the model
```{r}
# Loading an existing trained model
MOFAobject <- loadModel("BRCA_model1.hdf5")
MOFAobject
```

## Part 1: Disentangling the heterogeneity: calculation of variance explained by each factor in each view

The figure gives an overview of which latent factors (LF) are active in which view(s). If an LF is active in more than one view, this means that there is (co-variation) between features of different data modalities. Here, for example, LF1 is active in CNA and RNASeq, whwreas LF2 and LF3 are active in CNA and DNA methylation. 
```{r}
# Calculate the variance explained (R2) per factor in each view 
r2 <- calculateVarianceExplained(MOFAobject)
r2$R2Total
# Variance explained by each factor in each view
head(r2$R2PerFactor)
# Plot it
plotVarianceExplained(MOFAobject)
```

```{r}
fout <- plotFactorCor(MOFAobject)
```

## Part 2: Characterisation of individual factors
### Inspection of top weighted features in the active views
Inspection of top features with highest loadings: loading is a measure of feature importance, features with high loadings are the ones driving the heterogeneity captured by the factor. They provide the mapping between the high-dimensional space (the genes) and the low-dimensional space (the factors). They define a score for each gene on each factor, such that genes with no association with the factor are expected to have values close to zero, whereas genes with strong association with the factor are expected to have large absolute values.

#### Weight heat maps: They give an overview of the weights across all factors in a given view. 
```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "cna", 
  factors = 1:3,
  show_colnames = FALSE
)
```

```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "mut", 
  factors = 1:3,
  show_colnames = FALSE
)
```

```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "rnaseq", 
  factors = 1:3,
  show_colnames = FALSE
)
```

```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "meth", 
  factors = 1:3,
  show_colnames = FALSE
)
```

#### Weight plots: To explore a given factor in more detail we can plot all weights for a single factor.
```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 1, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 2, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 3, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "meth", 
  factor = 2, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "meth", 
  factor = 3, 
  nfeatures = 5
)
```


```{r}
plotWeights(
  MOFAobject, 
  view = "rnaseq", 
  factor = 1, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "mut", 
  factor = 3, 
  nfeatures = 5
)
```

If you are only interested in looking at only the top features you can use the `plotTopWeights` function.  
For example, here we plot the mutations with largest loadings on Factor 1. The sign on the right indicates the direction of the loading (positive/negative).
```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=1
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=2
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=3
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="meth", 
  factor=2
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="meth", 
  factor=3
)
```

Extracting the genomic information for the important CpGs based on the above plot

```{r}
a <- genomic_info["cg05422029",]
b <- genomic_info["cg16206460",]
c <- genomic_info["cg25976440",]
d <- genomic_info["cg24511738",]
e <- genomic_info["cg09165441",]
f <- genomic_info["cg18267049",]
g <- genomic_info["cg05937737",]
h <- genomic_info["cg22396555",]
i <- genomic_info["cg24604013",]
j <- genomic_info["cg03738025",]
rbind(a,b,c,d,e,f,g,h,i,j)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="rnaseq", 
  factor=1
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="mut", 
  factor=3
)
```


```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 1, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 2, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "meth", 
  factor = 2, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "meth", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "rnaseq", 
  factor = 1, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "mut", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```


# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```
