---
title: "Application of MOFA on BRCA data"
output: html_document
author: "Chaitra Rao"
date: "July 11, 2019"
---

```{r, cache = TRUE, echo=FALSE}
knitr::opts_chunk$set(error = TRUE, cache.lazy = FALSE)
```

## MOFA - Method Overview and its Application
Multi-Omics Factor Analysis (MOFA) is a statistical method used for integrating multiple modalities of omics data. It is an unsupervised integration method which learns a set of latent factors and aims to characterize heterogeneity among samples. These learnt latent factors (LFs) inferred by MOFA help capture the major sources of variation across these multiple modalities. They discover the major biological and technical sorces of variability. The factor loadings help in performing a broad range of downstream analyses, such as visualization, clustering and classification of samples in the low-dimensional space spanned by the factors. These factor loadings or weights are sparse, hence this facilitates linkage between factors and most relevant molecular features. 2 level regularization is required to ensure this sparsity: one encourages view and feature wise sparsity, meaning it directly identifies which factor is active in which view; second one encourages feature wise sparsity, which typically result in small number of features with active weights. The LFs can be unique to each dataset or be manifested across more than one dataset. This reveals the fact that there is a shared axes of variation between the different data modalities. The statistical framework that MOFA is based on consists of: 

* Fast inference based on a variational approximation
* Inference of sparse solutions facilitating interpretation
* Efficient handling of missing values
* Flexible combination of different likelihood models for each data modality, hence enabling integration of diverse data types, be it binary, count or continuous valued data.

MOFA was applied to a cohort of 1094 patient samples of BRCA, profiled for somatic mutations, CNA, RNASeq and DNA Methylation. This contained of diverse data types, hence the appropriate likelihood model was taken for each dataset:

* RNAseq data - continuous data - Gaussian likelihood
* DNA Methylation data - continuous data - Gaussian likelihood
* Somatic Mutation data - binary data - Bernoulli likelihood
* Copy Number Alteration data - continuous data - Gaussian likelihood

3 latent factors were identified, with a minimum explained variance set to 5%. This means that if a factor explained less than 5% of variance in any dataset, the factor would be dropped and the training would restart. This is known as the drop factor threshold which enables the pruning of factors during training. The convergence threshold for the model is based on the evidence lower bound (ELBO) and for this a value of 0.1 was used. To sum it up, the model was initialized with 25 factors and they were pruned using a threshold of variance explained of 5%. Further details on the data and training options can be found in the MOFA-BRCA.html or MOFA-BRCA.pdf file. 

Cumulatively, the 3 factors explained about 38% of variation in CNA data,  3% in DNA Methylation data, 2% in the RNAseq data and less than 1% in the somatic mutation data. This is visualized in the "Total Variance explained per view" plot.

Some important features were identified in BRCA which were aligned with specific factors. For this purpose the loadings of the features on individual factors were used to identify the etiology. Top weights plot was very useful for this purpose. For example, based on it in the CNA data, the individual LFs were found to be aligned with a set of genes with same abosulte loading. This could mean that the genes are most probably located close to each other and on the same chromosome.  I tried to connect LF1 factor to other molecular layer, namely, changes in the gene expression. We found that factor 1 was mainly associated with the gene expression status of ATP1A1 gene. LF1 captured about 17% variance in the CNA data, whereas in the gene expression data, it explained just about 2% variance. Another important factor was factor 2, active in both CNA and DNA methylation views. Analysis of the weights in the DNA methylation revealed that this factor tagged the CACNG7 gene, which is a calcium channel regulatory subunit. To check how this factor is associated with the gene expression data, I analysed its weights and it revealed that LF2 is aligned with gene expression of ARF1 gene. It has been previously established that ARF1 is the most amplified gene from the ARF gene family in breast cancer. [Xie. X et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5295416/pdf/oncotarget-07-58111.pdf) The LF3 was analysed for its role in CNA, DNA methylation and somatic mutation datasets. LF3 is aligned with the mutation status of the TP53 gene, whereas in the methylation dataset, the methylation status of ZIC5 gene is linked with LF3. About 9% of variance is captured by LF3 in the CNA dataset, 2% in the DNA methylation and less than 0.02% in the mutation dataset. 

## Outlook
With the help of MOFA, I was able to provide a basis which can serve as a guideline for further downstream analysis, such as gene set enrichment analysis, survival analysis. Gene set enrichment analysis, for example, would help in identifying the aetiologies of the top features related to certain biological pathways and processes. Furthermore, new likelihoods and noise models can be introduced which can encourage the application of MOFA in datasets with specific statistical properties. Using the latent factors to explain their associations with external data such as clinical covariates or use them as predictors in models of clinical outcomes would be an interesting approach as well. One of the limitations of MOFA is that the model is linear, in which case lot of non-linear relationships between features can be missed. Hence, a non-linear extension of MOFA may be able to tackle this problem, however one may expect some kind of trade-off between the model complexity, computational efficiency and interpretability. 

## Results
The results and the individual plots along with their siginificance are presented below. 

## Necessary Packages

```{r libraries, message = FALSE}
suppressPackageStartupMessages(library(RTCGAToolbox))
suppressPackageStartupMessages(library(TCGA2STAT))
suppressPackageStartupMessages(library(curatedTCGAData))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(lumi))
suppressPackageStartupMessages(library(MOFA))
suppressPackageStartupMessages(library(MultiAssayExperiment))
suppressPackageStartupMessages(library(survival))
```

Saving the genomic information of all the CpG sites.
```{r}
genomic_info <- meth.BRCA$cpgs
```

Characterictics of the model
```{r}
# Loading an existing trained model
MOFAobject <- loadModel("BRCA_model1.hdf5")
MOFAobject
```

## Part 1: Disentangling the heterogeneity: calculation of variance explained by each factor in each view

For each view, the coefficient of determination (R2) is calculated, which is the proportion of variance in the data explained by the latent facors (LFs) (both jointly or for each individual factor). For non-gaussian views (in our case, the somatic mutations which take binary data), the variance explained on the Gaussian pseudo-data is calculated. The plot gives an overview of which LFs are active in which view(s). In a way, it summarized the entire heterogeniety of the dataset. If an LF is active in more than one view, this means that there is (co-variation) between features of different data modalities. Here, for example, LF1 is active in CNA and RNASeq, whereas LF2 and LF3 are active in CNA and DNA methylation. Hence, one can conclude here that there is some co-variation between features of CNA and DNA Methylation explained by LF2 and LF3 and a co-variation is observed between CNA and RNAseq which is explained by LF1. This plot guides further characterization of the fcators by investigating the weights in those views. 
```{r}
# Calculate the variance explained (R2) per factor in each view 
r2 <- calculateVarianceExplained(MOFAobject)
r2$R2Total
# Variance explained by each factor in each view
head(r2$R2PerFactor)
# Plot it
plotVarianceExplained(MOFAobject)
```


#### Correlation matrix

A diagonal correlation matrix is expected because the model encourages the factors to be uncorrelated. A slight correlation might occur because it is not a hard constraint such as in PCA. But largely correlated factors are redundant and should be avoided to make the interpretation easier. In the below plot, one can see that it is mostly a diagonal correlation matrix, which slight correlations between the individual factors. However, these correlations are not too large, hence we can continue characterizing the individual factors.
```{r}
fout <- plotFactorCor(MOFAobject)
```

## Part 2: Characterisation of individual factors
### Inspection of top weighted features in the active views
Inspection of top features with highest loadings: One of the first steps for the annotation of a given factor is to visualize the corresponding loadings. Loading is a measure of feature importance, features with high loadings are the ones driving the heterogeneity captured by the factor. They provide the mapping between the high-dimensional space (the genes) and the low-dimensional space (the factors). They define a score for each gene on each factor, such that genes with no association with the factor are expected to have values close to zero, whereas genes with strong association with the factor are expected to have large absolute values. Important plots here are the weight heatmaps, weight plots, top weight plots and data heatmaps.

#### Weight Heatmaps: They give an overview of the weights across all factors in a given view. 
Visualize the overall pattern of weights, but not characterize the individual factors. 
```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "cna", 
  factors = 1:3,
  show_colnames = FALSE
)
```
Here, we observe that all the factors have large non-zero values. Hence, it would be interesting to investigate these factors individually. 
```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "mut", 
  factors = 1:3,
  show_colnames = FALSE
)
```
In the "variance explained per factor" plot, we saw that none of the factors explained a lot of variation in somatic mutations. However, in the above plot, we can see that LF1 might be explaining some variation as it has slightly large non-zero value, but also LF2 and LF3 to some extent.
```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "rnaseq", 
  factors = 1:3,
  show_colnames = FALSE
)
```
In case of RNAseq, LF1 has larger non-zero values compared to remaining LFs.
```{r}
plotWeightsHeatmap(
  MOFAobject, 
  view = "meth", 
  factors = 1:3,
  show_colnames = FALSE
)
```
As we saw in the "variance explained per factor" plot as well, LF2 and LF3 are active in DNA Methylation view and hence they also have large non-zero values, as seen in the weight heatmaps. Thus, we analyze these factors further in more detail.

#### Weight plots: To explore a given factor in more detail we can plot all weights for a single factor.
```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 1, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 2, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "cna", 
  factor = 3, 
  nfeatures = 5
)
```
Features with large absolute weight on a given factor explain a pattern of covariation associated with that particular factor. The reason that the factors are aligned with so many features with similar weight loadings could be that the genes are very close to each other. The variance decomposition plots reveal that there is a shared axes of variation between the different omics layers. Hence, LF1 will be further characterized for its role in RNAseq view, LF2 in the DNA methylation view and LF3 in DNA methylation and mutation views.
```{r}
plotWeights(
  MOFAobject, 
  view = "meth", 
  factor = 2, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "meth", 
  factor = 3, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "rnaseq", 
  factor = 1, 
  nfeatures = 5
)
```

```{r}
plotWeights(
  MOFAobject, 
  view = "mut", 
  factor = 3, 
  nfeatures = 5
)
```


#### Top Weight plots: Plots featuring top weights for a given LF in a given view
With these plots, we can only look at the top features. The sign on the right indicates the direction of the loading (positive/negative). A positive loading indicates that the feature is more active in the cells with positive factor values, while a negative loading indicates that the feature is more active in the cells with negative factor values.
```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=1
)
```


```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=2
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="cna", 
  factor=3
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="meth", 
  factor=2
)
```

Extracting the genomic information for the important CpGs based on the above plot.
```{r}
a <- genomic_info["cg21477176",]
b <- genomic_info["cg01851378",]
c <- genomic_info["cg04118306",]
d <- genomic_info["cg11213690",]
e <- genomic_info["cg14231297",]
f <- genomic_info["cg26995244",]
g <- genomic_info["cg17078116",]
h <- genomic_info["cg26132320",]
i <- genomic_info["cg05422029",]
j <- genomic_info["cg25976440",]
rbind(a,b,c,d,e,f,g,h,i,j)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="meth", 
  factor=3
)
```

```{r}
a <- genomic_info["cg18557131",]
b <- genomic_info["cg20395967",]
c <- genomic_info["cg03738025",]
d <- genomic_info["cg26995244",]
e <- genomic_info["cg26444528",]
f <- genomic_info["cg11213690",]
g <- genomic_info["cg26132320",]
h <- genomic_info["cg24604013",]
i <- genomic_info["cg05937737",]
j <- genomic_info["cg22396555",]
rbind(a,b,c,d,e,f,g,h,i,j)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="rnaseq", 
  factor=1
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="rnaseq", 
  factor=2
)
```

```{r}
plotTopWeights(
  MOFAobject, 
  view="mut", 
  factor=3
)
```


#### Data Heatmaps: Heatmaps featuring the input data for relevant features
Instead of looking at an "abstract" weight, it is useful to observe the coordinated heterogeneity of the top features in the original data. In this plot samples (in rows) are ordered according to their value on the factor. This function extracts the top features for a given factor and view, and generates a heatmap with dimensions (samples,features). This should reveal the underlying heterogeneity that is captured by the latent factor.
```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 1, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 2, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "cna", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "meth", 
  factor = 2, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "meth", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "rnaseq", 
  factor = 1, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "rnaseq", 
  factor = 2, 
  features = 5, 
  show_rownames = FALSE
)
```

```{r}
plotDataHeatmap(
  MOFAobject, 
  view = "mut", 
  factor = 3, 
  features = 5, 
  show_rownames = FALSE
)
```

This function extracts the weights from the model for all views and factors.
```{r}
getWeights(MOFAobject, views = "all", factors = "all",
  as.data.frame = TRUE)

```


# Session info

```{r sessionInfo, cache = TRUE}
date()
devtools::session_info()
```
